### PR TITLE
added state mutex to Promise

### PR DESCRIPTION
**Pull request by sebastianbaginski**
Fixes race conditions on `_state` reported by helgrind.
@marcusnaslund

### CODE REVIEW

**Comment by marcusnaslund** [comment]
Let's have @emilwestergren weigh in here too.

**Comment by sebastianbaginski** [comment]
@marcusnaslund looks like PRInt machine has some problems with local settings file

**Comment by marcusnaslund** [comment]
@sebastianbaginski Oops.

**Comment by marcusnaslund** [comment]
Now it works.

**Comment by sebastianbaginski** [comment]
Seems to be stuck for ~1h now : `09:29:22 SomeTest ...`

**Comment by marcusnaslund** [comment]
“He that can have patience can have what he will.”
― Benjamin Franklin

**Comment by marcusnaslund** [comment]
What happened here?

**Comment by sebastianbaginski** [comment]
I deleted the branch locally and pushed the changes upstream.

**Comment by marcusnaslund** [comment]
Less work for me to do. :)

**Comment by emilwestergren** [comment]
You can use `mutex with(|| ...)` if you want to get rid of some lines.

**Comment by marcusnaslund** [comment]
Let's hope this fixes #1728 too.

File: source/concurrent/Promise.ooc [hunk 1]
```diff
@@ -131,19 +156,27 @@ _ThreadFuture: class <T> extends Future<T> {
...
 			if (this _thread wait(time toSeconds()))
 				this _threadAlive = false
-		this _state == _PromiseState Finished
+		this _stateMutex lock()
+		result := this _state == _PromiseState Finished
+		this _stateMutex unlock()
+		result
 	}
 	getResult: override func (defaultValue: T) -> T {
 		result := defaultValue
```
Comment (marcusnaslund) [inline source/concurrent/Promise.ooc]: Does this line outside the lock also suffer problems?

Comment (sebastianbaginski) [inline source/concurrent/Promise.ooc]: No, looks like it's just a memcpy with a constant size:
```ooc
Test: class <T> {
    init: func
    fun: func (value: T) {
		x := value
    }
}


void prob__Test_fun(prob__Test* this, uint8_t* value) {
    uint8_t* x = Memory__alloca(this->T->size);
    Memory__memcpy(x, value, this->T->size);
}
```
no "class constructor" calls here.

Comment (marcusnaslund) [inline source/concurrent/Promise.ooc]: Thanks rock.

**Pull request merged**